### DISCUSSION TITLE
Feature to merge `__table_args__` from inherited mixins

### DISCUSSION THREAD

**Discussion by pykenny** [sqlalchemy/sqlalchemy#8911] [Ideas]
Recently I'm trying to combine the two mechanisms below for defining table arguments within ORM model. Lots of the tables/models I'm building are sharing some common patterns, so I believe using mixins can greatly reduce redundancies:

- [Defining index in mixins](https://docs.sqlalchemy.org/en/14/orm/declarative_mixins.html#creating-indexes-with-mixins), and
- [Merging table arguments from inherited mixins](https://docs.sqlalchemy.org/en/14/orm/declarative_mixins.html#combining-table-mapper-arguments-from-multiple-mixins)

Things are relatively simple when we're sure all the mixins defines their `__table_args__` with a dictionary:
```python
# Within a model class
@declared_attr
def __table_args__(cls):
    return {
        **MixinOne.__table_args__,
        **MixinTwo.__table_args__,
        **MixinThree.__table_args__,
    }
```

Would be a bit redundant when lots of mixins involved, but given the readability it's bearable.

However, things get more complicated when mixin-defined index gets involved, since now `__table_args__` can be one of the formats below:
 - A tuple with schema constructs
 - A tuple with schema constructs, except the last one as a dictionary
 - A dictionary

And this brings me to some embarrassing situation where I find myself need to paste these logic in all the models sharing these mixins, which greatly reduces code readability:
```python
# Within a model class
@declared_attr
def __table_args__(cls):
    args = []
    kwargs = {}
    for mixin in (MixinOne, MixinTwo, MixinThree):  # <-- ((Customizable))
        table_args = getattr(mixin, "__table_args__", None)
        if table_args:  # Not None, or having at least one member in the collection
            if isinstance(table_args, tuple):
                last_positional_arg = table_args[-1]
                args.extend(table_args[:-1])
                if isinstance(last_positional_arg, dict):
                    kwargs.update(last_positional_arg)
                else:
                    args.append(last_positional_arg)
            elif isinstance(table_args, dict):
                kwargs.update(table_args)

    # Place my positional arguments in args.
    ... # <-- ((Customizable))
    # Also place my keyword arguments in kwargs
    ... # <-- ((Customizable))

    return ((*args, kwargs) if kwargs else tuple(args)) if args else (kwargs or ())
```

And up until now I really want a function to relieve such burden 😂:  
```python
from somewhere import merge_table_arguments

# Within a model class
@declared_attr
def __table_args__(cls):
    return merge_table_arguments(MixinOne, MixinTwo, MixinThree, additions={"schema": "my_database"})
```

**Comment by zzzeek** [comment]
hi -

you could take that code you wrote and stick it in your own merge_table_arguments() function, and put that in your util.py somewhere.    I would gladly put it on the [wiki](https://github.com/sqlalchemy/sqlalchemy/wiki/UsageRecipes) as well.    

Once I add something like that as a "feature", now it's on me to make it do everything everyone might want.   such as what if two of your mixins had conflicting or repeated constructs in `(args, )`, things like that.  Having people roll this themselves lets them take responsibility for all the weird things they might need to have it do.

**Comment by pykenny** [comment]
Finally come up with this solution, after considering the [need for accessing final class' attributes](https://github.com/sqlalchemy/sqlalchemy/discussions/8914):

```python
def merge_table_arguments(cls, *mixins, yours = None):
    args = []
    kwargs = {}

    table_args_theirs = (
        getattr(super(base_cls, cls), "__table_args__", None)
        for base_cls in (cls, *mixins)
    )
    
    for table_args in (*table_args_theirs, yours):
        if table_args:
            if isinstance(table_args, tuple):
                last_positional_arg = table_args[-1]
                args.extend(table_args[:-1])
                if isinstance(last_positional_arg, dict):
                    kwargs.update(last_positional_arg)
                else:
                    args.append(last_positional_arg)
            elif isinstance(table_args, dict):
                kwargs.update(table_args)
    
    if args:
        if kwrags:
            return (*args, kwargs)
        else:
            return tuple(args)
    else:
        return kwargs
```

Hence we'll be able to write something like this:
```python
class MyModel(MixinOne, MixinTwo, MixinThree, Base):
    __tablename__ = "my_table"

    @ declared_attr
    def __table_args__(cls):
        return merge_table_arguments(cls, MixinOne, MixinTwo, MixinThree, yours={"schema": "my_database"})

    ...
```

**Reply by CaselIT** [reply]
Maybe we could add a wiki article for this. Thoughts @zzzeek ?

**Reply by zzzeek** [reply]
it needs to have a complete, tested example illustrating at least two mixins and some arguments being merged for it to make sense to people, otherwise sure!

**Reply by pykenny** [reply]
One issue where the merge function will not work as expected is repeated construct, as @zzzeek mentioned earlier. Given both mixin `A` and `B` inherit mixin `C` and call `merge_table_arguments` to combine table arguments of `C`, where `C` itself defines an index as positional argument in `__table_args__`. When a model class `D` inherits `A` and `B`, then calling `merge_table_arguments` inside of `__table_args__`'s implementation will lead to three duplicated indexes in final outcome, since:
 - `A` defines the index in `__table_args__`
 - `B` defines the index in `__table_args__`
 - Calling `super()` on one of `A` or `B` will lead to `C` as Python's naming resolution.

**Reply by pykenny** [reply]
One way to avoid this problem is to make everything flattened (which means mixins doesn't inherit each other), and also make sure every positional argument in these mixins are exclusive. Keyword arguments are totally fine as long as we can overwrite with the correct one in `yours` argument when calling the function, but we lose that kind of luck in positional arguments.

Right now the use case I'm working on is relatively simple so that's entirely feasible, but I'm not sure about how, how often, and how complicated other developers use these declarative features in production.

**Comment by notatallshaw-gts** [comment]
Hi all, I recently came across this problem and wrote my own helper function that for my use case is simpler to use, it does the following things different:

* No manually passing in parent models in the class method
* Adds positional arguments in reverse parent order
* Requires parent classes to explicitly merge with their own parents or not
* Does not allow duplicate positional arguments

I make no guarantees this works, in particular I do not have a strong understanding of the wider use of SqlAlchemy so if you use please test thoroughly for your use case:

```python
def table_args(
        cls: type,
        __table_args__: dict[str, Any] | tuple[Any, ...]
    ) -> dict[str, Any] | tuple[Any, ...]:
    """
    Helper class function to merge the __table_args__ attr of a
    SQLAlchemy model that has some inheritance. Use it like this:

    from sqlalchemy.ext.declarative import declared_attr

    class MyModel(MixinOne, MixinTwo, MixinThree, Base):
        __tablename__ = "my_table"

        @declared_attr
        def __table_args__(cls):
            return table_args(cls, {"schema": "my_database"})
    """
    args: dict[Any, None] = {}  # Store args in keys of dictionary for fast lookup and no duplicates
    kwargs: dict[str, Any] = {}
    table_args_attrs = []
    
    # Collect immediate parent "__table_args__" in reverse order
    for class_ in reversed(cls.__bases__):   
        table_args_attr = getattr(class_, "__table_args__", None)
        if table_args_attr:
            table_args_attrs.append(table_args_attr)

    # Also merge the current table args
    table_args_attrs.append(__table_args__)

    for table_args_attr in table_args_attrs:
        # Handle simple dictionary use case first
        if isinstance(table_args_attr, dict):
            kwargs.update(table_args_attr)
            continue

        if isinstance(table_args_attr, tuple):
            if not table_args_attr:
                raise ValueError(f'Empty table arguments found')
            
            # Add all but last positional argument, if it is not already added
            for positional_arg in table_args_attr[:-1]:
                if positional_arg not in args:
                    args[positional_arg] = None

            last_positional_arg = table_args_attr[-1]
            if isinstance(last_positional_arg, dict):
                kwargs.update(last_positional_arg)
            elif last_positional_arg not in args:
                args[last_positional_arg] = None
    
    # Either return a tuple or a dictionary depending on what was found
    if args:
        if kwargs:
            return (*args, kwargs)
        else:
            return tuple(args)
    else:
        return kwargs
```
